spec/controllers/api/v1/punches_controller_spec.rb

```diff
@@ -0,0 +1,155 @@
+require 'spec_helper'
+
+describe Api::V1::PunchesController, :type => :controller do
+  let(:company) { create(:company) }
+  let(:user) { create(:user, :with_token, company: company) }
+  let(:project) { create(:project, company: user.company) }
+  let(:headers) { { token: user.token } }
+
+  describe 'POST api/v1/punches' do
+    context 'when api token is valid and params are valid' do
+      let(:params) {
+        {
+          "_json": [
```

Comment:
You could call this key `punches` for better readability.

Reply:
When I send a request with the params:
```json
[
  {
    "from": "2022-04-19T12:00:00.000Z",
    "to": "2022-04-19T15:00:00.000Z",
    "project_id": "3"
  },
  {
    "from": "2022-04-19T17:00:00.000Z",
    "to": "2022-04-19T21:00:00.000Z",
    "project_id": "3"
  }
]
```
The rails put its values on a params called _json, but when I send by test, I need to declare this key

Reply:
I see, perhaps that's the default behavior when you send a JSON array instead of an object.

But what I meant is to make the payload expect a `punches` key, so that the payload would be something like:
```json
{
  "punches": {
    "from": "2022-04-19T12:00:00.000Z",
    "to": "2022-04-19T15:00:00.000Z",
    "project_id": "3"
  },
  {
    "from": "2022-04-19T17:00:00.000Z",
    "to": "2022-04-19T21:00:00.000Z",
    "project_id": "3"
  }
}
```

You could then load the data from inside the controller by using:
```rb
def punch_params
  params.require(:punches).permit(punches: [:from, :to, :project_id])
end
```

Reply:
Yes, I did it, but the require must be after permit, because the require returns an array and does not have the permit method on array